languages/rust/src/query.rs

```diff
@@ -0,0 +1,282 @@
+    }
+
+    fn handle_external_call(
+        &mut self,
+        call_id: u64,
+        instance: Term,
+        name: Name,
+        args: Option<Vec<Term>>,
+    ) -> anyhow::Result<()> {
+        let instance = Instance::from_polar(&instance, &mut self.host.lock().unwrap()).unwrap();
+        if let Err(e) = self.register_call(call_id, instance, name, args) {
+            self.application_error(e);
+            return self.call_result_none(call_id);
+        }
+
+        if let Some(result) = self.next_call_result(call_id) {
+            self.call_result(call_id, result)
+        } else {
+            self.call_result_none(call_id)
+        }
+    }
+
+    fn handle_external_op(
+        &mut self,
+        call_id: u64,
+        operator: Operator,
+        args: Vec<Term>,
+    ) -> anyhow::Result<()> {
+        assert_eq!(args.len(), 2);
+        let res = {
+            let mut host = self.host.lock().unwrap();
+            let args = [
+                Instance::from_polar(&args[0], &mut host).unwrap(),
+                Instance::from_polar(&args[1], &mut host).unwrap(),
+            ];
+            host.operator(operator, args)
+        };
+        self.question_result(call_id, res);
+        Ok(())
+    }
+
+    fn handle_external_isa(
+        &mut self,
+        call_id: u64,
+        instance: Term,
+        class_tag: Name,
+    ) -> anyhow::Result<()> {
+        tracing::debug!(instance = ?instance, class = %class_tag, "isa");
+        let res = self.host.lock().unwrap().isa(instance, &class_tag);
+        self.question_result(call_id, res);
+        Ok(())
+    }
+
+    fn handle_external_unify(
+        &mut self,
+        call_id: u64,
+        left_instance_id: u64,
+        right_instance_id: u64,
+    ) -> anyhow::Result<()> {
+        let res = self
+            .host
+            .lock()
+            .unwrap()
+            .unify(left_instance_id, right_instance_id);
+        self.question_result(call_id, res);
+        Ok(())
+    }
+
+    fn handle_external_is_subspecializer(
+        &mut self,
+        call_id: u64,
+        instance_id: u64,
+        left_class_tag: Name,
+        right_class_tag: Name,
+    ) -> anyhow::Result<()> {
+        let res = self.host.lock().unwrap().is_subspecializer(
+            instance_id,
+            &left_class_tag,
+            &right_class_tag,
+        );
+        self.question_result(call_id, res);
+        Ok(())
+    }
+
+    fn handle_debug(&mut self, message: String) -> anyhow::Result<()> {
+        eprintln!("TODO: {}", message);
+        self.check_messages();
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+pub struct ResultSet {
+    pub bindings: polar_core::types::Bindings,
+    pub host: Arc<Mutex<crate::host::Host>>,
+}
+
+impl ResultSet {
+    pub fn get<T: crate::host::FromPolar>(&self, name: &str) -> Option<T> {
+        self.bindings
+            .get(&Name(name.to_string()))
+            .and_then(|term| T::from_polar(term, &mut self.host.lock().unwrap()))
```

Comment:
What happens in this case:

```
# policy.polar

allow(1);
allow("1");
allow(1.1);
allow([1, 2]);
```

```
// test.rs

fn test_policy() {
    let oso = Oso::new();
    // I don't think we have a Variable type right now, anyway...
    let query = oso.query_rule("allow", Variable("x"));

    for result in query {
        let x = result.get("x");
        println("{}", x);
    }
}
```

The above will fail to compile because rust won't know the type of `T`.

```
// test_typed.rs

fn test_policy() {
    let oso = Oso::new();
    // I don't think we have a Variable type right now, anyway...
    let query = oso.query_rule("allow", Variable("x"));

    for result in query {
        let x: u64 = result.get("x");
        println("{}", x);
    }
}
```

This will succeed for the first rule, but return `None` for the rest of them.

Since we are going from dynamic to static types here, I think we need a way to have `result.get()` return an enum so that the user can have rules that potentially return multiple types, then match and extract the value as they want.

I also think we need to return an Error type from `get()` so that we can distinguish between the absence of a value (i.e no binding named "x") and an incorrect type (there is a binding named "x", but it cannot be converted to the result type `T`).

Reply:
Easy one first:

> I also think we need to return an Error type from get() so that we can distinguish between the absence of a value

💯 

> Since we are going from dynamic to static types here, I think we need a way to have result.get() return an enum so that the user can have rules that potentially return multiple types, then match and extract the value as they want.

Yeah @saolsen raised the same issue. Our plan was to have an option (maybe even the default?) be to just return the `Value` enum. You can do what you want with it. But also provide a helper method, like what we have here, that also does the conversion for you. What do you think?